File: backend/service-b/src/report/report.service.ts
import {
  BadRequestException,
  Injectable,
  Logger,
  OnModuleInit,
  ServiceUnavailableException,
} from '@nestjs/common';
import * as fs from 'fs';
import * as grpc from '@grpc/grpc-js';
import * as protoLoader from '@grpc/proto-loader';
import * as path from 'path';
import { MetricPoint, TelemetryService } from '@fs-assignment/common';

const GRPC_DEADLINE_MS = 15_000;
const HOUR_MS = 60 * 60 * 1000;
const DAY_MS = 24 * HOUR_MS;

interface ReportGrpcClient {
  GeneratePdf: (
    request: { filter_date: string; filter_type: string; time_series_json: Buffer },
    options: { deadline: Date },
    callback: (error: grpc.ServiceError | null, response: { pdf_content: Buffer }) => void,
  ) => void;
}

@Injectable()
export class ReportService implements OnModuleInit {
  private readonly logger = new Logger(ReportService.name);
  private client!: ReportGrpcClient;

  constructor(private readonly telemetry: TelemetryService) {}

  onModuleInit() {
    const packageDefinition = protoLoader.loadSync(this.resolveProtoPath(), {
      keepCase: true,
      longs: String,
      enums: String,
      defaults: true,
      oneofs: true,
    });
    const reportProto = grpc.loadPackageDefinition(packageDefinition) as grpc.GrpcObject;
    const reportPackage = reportProto.report as { ReportService: grpc.ServiceClientConstructor };
    const host = process.env.REPORT_GRPC_HOST || 'localhost';
    const port = process.env.REPORT_GRPC_PORT || '50051';
    this.client = new reportPackage.ReportService(
      `${host}:${port}`,
      grpc.credentials.createInsecure(),
    ) as unknown as ReportGrpcClient;
    this.logger.log(`Report gRPC client -> ${host}:${port}`);
  }

  /**
   * Builds a PDF for the requested day (UTC) from real time series data.
   * If metrics cannot be read, the request fails — the report never renders
   * invented numbers.
   */
  async generatePdf(date: string, type?: string): Promise<Buffer> {
    const from = Date.parse(`${date}T00:00:00.000Z`);
    if (Number.isNaN(from)) {
      throw new BadRequestException(`Invalid date: ${date}`);
    }
    const to = from + DAY_MS - 1;
    const actionFilter = type && type !== 'all' ? type : undefined;

    let points: MetricPoint[];
    try {
      points = await this.telemetry.getMetricsRange(from, to, HOUR_MS, actionFilter);
    } catch (err) {
      this.logger.error(`Metrics read failed: ${(err as Error).message}`);
      throw new ServiceUnavailableException('Metrics store is unavailable, report cannot be generated');
    }

    const timeSeriesJson = Buffer.from(JSON.stringify(points));
    return new Promise((resolve, reject) => {
      this.client.GeneratePdf(
        { filter_date: date, filter_type: actionFilter ?? 'all', time_series_json: timeSeriesJson },
        { deadline: new Date(Date.now() + GRPC_DEADLINE_MS) },
        (error, response) => {
          if (error) {
            this.logger.error(`gRPC GeneratePdf failed: ${error.message}`);
            reject(new ServiceUnavailableException('PDF service is unavailable'));
          } else {
            resolve(response.pdf_content);
          }
        },
      );
    });
  }

  private resolveProtoPath(): string {
    const candidates = [
      path.join(process.cwd(), 'proto', 'report.proto'),
      path.join(process.cwd(), '..', 'proto', 'report.proto'),
      path.join(__dirname, '..', '..', 'proto', 'report.proto'),
    ];
    const found = candidates.find((candidate) => fs.existsSync(candidate));
    if (!found) {
      throw new Error(`report.proto not found. Checked: ${candidates.join('; ')}`);
    }
    return found;
  }
}

File: backend/service-b/internal/service/report_service.go
package service

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"time"

	"github.com/fs-assignment/service-b/internal/repository"
)

const hourMs = int64(time.Hour / time.Millisecond)
const dayMs = 24 * hourMs

var ErrMetricsUnavailable = errors.New("metrics store is unavailable, report cannot be generated")

var ErrPdfServiceUnavailable = errors.New("PDF service is unavailable")

type PDFGenerator interface {
	GeneratePdf(ctx context.Context, filterDate, filterType string, timeSeriesJSON []byte) ([]byte, error)
}

type ReportService struct {
	metrics repository.MetricsRepository
	pdf     PDFGenerator
}

func NewReportService(metrics repository.MetricsRepository, pdf PDFGenerator) *ReportService {
	return &ReportService{metrics: metrics, pdf: pdf}
}

func (s *ReportService) GeneratePdf(ctx context.Context, date, actionType string) ([]byte, error) {
	from, err := time.Parse("2006-01-02", date)
	if err != nil {
		return nil, fmt.Errorf("invalid date: %s", date)
	}
	fromMs := from.UTC().UnixMilli()
	toMs := fromMs + dayMs - 1

	actionFilter := actionType
	if actionFilter == "all" {
		actionFilter = ""
	}

	points, err := s.metrics.GetRange(ctx, fromMs, toMs, hourMs, actionFilter)
	if err != nil {
		return nil, ErrMetricsUnavailable
	}

	timeSeriesJSON, err := json.Marshal(points)
	if err != nil {
		return nil, err
	}

	filterType := actionFilter
	if filterType == "" {
		filterType = "all"
	}

	pdf, err := s.pdf.GeneratePdf(ctx, date, filterType, timeSeriesJSON)
	if err != nil {
		return nil, ErrPdfServiceUnavailable
	}
	return pdf, nil
}
